src/InjectCoordinatesIntrinsics.cpp

```diff
@@ -0,0 +1,156 @@
+using std::vector;
+
+class InjectCoordinatesIntrinsics : public IRMutator {
+public:
+    InjectCoordinatesIntrinsics() : inside_kernel_loop(false) {}
+    Scope<int> scope;
+    bool inside_kernel_loop;
+
+private:
+    using IRMutator::visit;
+
+    void visit(const Provide *provide) {
+        if (!inside_kernel_loop) {
+            IRMutator::visit(provide);
+            return;
+        }
+
+        internal_assert(provide->values.size() == 1)
+            << "Coordinate currently only supports single-valued stores.\n";
+        user_assert(provide->args.size() == 3)
+            << "Coordinate stores require three coordinates.\n";
+
+        // Create coordinate_store("name", name.buffer, x, y, c, value)
+        // intrinsic.
+        Expr value_arg = mutate(provide->values[0]);
+        vector<Expr> args = {
+            provide->name,
+            Variable::make(Handle(), provide->name + ".buffer"),
+            provide->args[0],
+            provide->args[1],
+            provide->args[2],
+            value_arg};
+
+        stmt = Evaluate::make(Call::make(value_arg.type(),
+                                         Call::coordinates_store,
+                                         args,
+                                         Call::Intrinsic));
+    }
+
+    void visit(const Call *call) {
+        if (!inside_kernel_loop || call->call_type == Call::Intrinsic ||
+            call->call_type == Call::Extern) {
+            IRMutator::visit(call);
+            return;
+        }
+
+        string name = call->name;
+        if (call->call_type == Call::Halide && call->func.outputs() > 1) {
+            name = name + '.' + int_to_string(call->value_index);
+        }
+
+        vector<Expr> call_args = call->args;
+        // Check to see if we are reading from a one or two dimension function
+        // and pad to three dimensions.
+        while (call_args.size() < 3) {
+            call_args.push_back(IntImm::make(0));
+        }
+
+        // Create coordinates_load("name", "name[.n]", name.buffer, x, y, c)
+        // intrinsic call.
+        // We need to pass "name[.n]" because if we need to add normalization
+        // then we will this name as a prefix for "name[.n].extent" variable.
+        vector<Expr> args(6);
+        args[0] = call->name;
+        args[1] = name;
+        args[2] = Variable::make(Handle(), call->name + ".buffer");
+        for (size_t i = 0; i < call_args.size(); i++) {
+            string d = int_to_string(i);
+            string min_name = name + ".min." + d;
+            string min_name_constrained = min_name + ".constrained";
+            if (scope.contains(min_name_constrained)) {
+                min_name = min_name_constrained;
+            }
+
+            Expr min = Variable::make(Int(32), min_name);
+
+            // Remind users to explicitly specify the 'min' values of
+            // ImageParams accessed by coordinate-based filters.
+            if (i == 2 && call->param.defined()) {
```

Comment:
You get an undefined min if you never call set_min

I think it should be a user_error. You're only breaking code that is
already silently broken but happens to work because people only ever pass
images that start at zero.

On Tue, Apr 14, 2015 at 8:58 PM, Alexander Aprelev <notifications@github.com
> wrote:

> In src/InjectCoordinatesIntrinsics.cpp
> <https://github.com/halide/Halide/pull/746#discussion_r28391941>:
>
> > +        args[0] = call->name;
> > +        args[1] = name;
> > +        args[2] = Variable::make(Handle(), call->name + ".buffer");
> > +        for (size_t i = 0; i < call_args.size(); i++) {
> > +            string d = int_to_string(i);
> > +            string min_name = name + ".min." + d;
> > +            string min_name_constrained = min_name + ".constrained";
> > +            if (scope.contains(min_name_constrained)) {
> > +                min_name = min_name_constrained;
> > +            }
> > +
> > +            Expr min = Variable::make(Int(32), min_name);
> > +
> > +            // Remind users to explicitly specify the 'min' values of
> > +            // ImageParams accessed by coordinate-based filters.
> > +            if (i == 2 && call->param.defined()) {
>
> I moved original logic of generating warning and defaulting to min equals
> to 0 back to InjectOpenGLIntrinsics. I am not sure if I change that(to
> user_assert?) whether it will break somebody. More basic question though -
> how one can get an non-constant or undefined min for an image? set_min,
> set_bounds take ints as arguments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/746/files#r28391941>.
>